src/specFunctions.py

```diff
@@ -30,413 +30,426 @@
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    THE SOFTWARE.
+    '''
+    if isinstance(y, np.ndarray) and np.issubdtype(y.dtype, np.unsignedinteger):
+        raise ValueError("y must be signed")
+
+    if not thres_abs:
+        thres = thres * (np.max(y) - np.min(y)) + np.min(y)
+
+    min_dist = int(min_dist)
+
+    # compute first order difference
+    dy = np.diff(y)
+
+    # propagate left and right values successively to fill all plateau pixels (0-value)
+    zeros, = np.where(dy == 0)
+
+    # check if the signal is totally flat
+    if len(zeros) == len(y) - 1:
+        return np.array([])
+
+    if len(zeros):
+        # compute first order difference of zero indexes
+        zeros_diff = np.diff(zeros)
+        # check when zeros are not chained together
+        zeros_diff_not_one, = np.add(np.where(zeros_diff != 1), 1)
+        # make an array of the chained zero indexes
+        zero_plateaus = np.split(zeros, zeros_diff_not_one)
+
+        # fix if leftmost value in dy is zero
+        if zero_plateaus[0][0] == 0:
+            dy[zero_plateaus[0]] = dy[zero_plateaus[0][-1] + 1]
+            zero_plateaus.pop(0)
+
+        # fix if rightmost value of dy is zero
+        if len(zero_plateaus) and zero_plateaus[-1][-1] == len(dy) - 1:
+            dy[zero_plateaus[-1]] = dy[zero_plateaus[-1][0] - 1]
+            zero_plateaus.pop(-1)
+
+        # for each chain of zero indexes
+        for plateau in zero_plateaus:
+            median = np.median(plateau)
+            # set leftmost values to leftmost non zero values
+            dy[plateau[plateau < median]] = dy[plateau[0] - 1]
+            # set rightmost and middle values to rightmost non zero values
+            dy[plateau[plateau >= median]] = dy[plateau[-1] + 1]
+
+    # find the peaks by using the first order difference
+    peaks = np.where(
+        (np.hstack([dy, 0.0]) < 0.0)
+        & (np.hstack([0.0, dy]) > 0.0)
+        & (np.greater(y, thres))
+    )[0]
+
+    # handle multiple peaks, respecting the minimum distance
+    if peaks.size > 1 and min_dist > 1:
+        highest = peaks[np.argsort(y[peaks])][::-1]
+        rem = np.ones(y.size, dtype=bool)
+        rem[peaks] = False
+
+        for peak in highest:
+            if not rem[peak]:
+                sl = slice(max(0, peak - min_dist), peak + min_dist + 1)
+                rem[sl] = True
+                rem[peak] = False
+
+        peaks = np.arange(y.size)[~rem]
+
+    return peaks    
+
+def compute_wavelengths(width, pixels, wavelengths, errors):
```

Comment:
Splitting `readcal` in two. This way we can call the function separately elsewhere.

I use this when calibrating a set of points so that I can output the error of the current selection. That way I can choose to update the caldata file or undo my changes if I've managed to do worse than my previous calibration.